core/internal/filetransfer/file_transfer_s3.go

```diff
@@ -0,0 +1,262 @@
+		if err != nil {
+			return ft.formatDownloadError("error constructing object input", err)
+		}
+		getObjectInputs = []*s3.GetObjectInput{getObjInput}
+	} else {
+		getObjectInputs, err = ft.listObjectsWithPrefix(bucketName, rootObjectName)
+		if err != nil {
+			return ft.formatDownloadError(
+				fmt.Sprintf(
+					"error getting objects in bucket %s under prefix %s",
+					bucketName, rootObjectName,
+				),
+				err,
+			)
+		}
+	}
+	err = ft.downloadFiles(rootObjectName, getObjectInputs, task.PathOrPrefix)
+	if err != nil {
+		return ft.formatDownloadError("error downloading object", err)
+	}
+	return nil
+}
+
+// findObjectFromTask finds the s3 object that matches the versionId and
+// digest/ETag specified by the task, and returns a struct that we can use
+// to access that s3 object.
+func (ft *S3FileTransfer) findObjectFromTask(
+	bucketName string,
+	objectName string,
+	task *ReferenceArtifactDownloadTask,
+) (*s3.GetObjectInput, error) {
+	var getObjInput *s3.GetObjectInput = &s3.GetObjectInput{
+		Bucket: aws.String(bucketName),
+		Key:    aws.String(objectName),
+	}
+	var getObjAttrsInput *s3.GetObjectAttributesInput = &s3.GetObjectAttributesInput{
+		Bucket:           aws.String(bucketName),
+		Key:              aws.String(objectName),
+		ObjectAttributes: []types.ObjectAttributes{types.ObjectAttributesEtag},
+	}
+
+	versionId, ok := task.VersionIDString()
+	if ok {
+		getObjAttrsInput.VersionId = &versionId
+		getObjInput.VersionId = &versionId
+	}
+
+	objAttrs, err := ft.client.GetObjectAttributes(ft.ctx, getObjAttrsInput)
+	if err != nil {
+		return nil, err
+	}
+
+	// If the ETag doesn't match what we have stored, try to find the correct version
+	if strings.Trim(*objAttrs.ETag, "\"") != task.Digest {
+		if task.VersionId != nil {
+			return nil, fmt.Errorf(
+				"digest/etag mismatch: etag %s does not match expected digest %s",
+				*objAttrs.ETag,
+				task.Digest,
+			)
+		}
+		getObjInput, err = ft.getCorrectObjectVersion(
+			bucketName,
+			objectName,
+			task.Digest,
+			getObjInput,
+		)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return getObjInput, nil
+}
+
+// listObjectsWithPrefix returns a list of all objects in the specified bucket
+// that begin with the given prefix.
+func (ft *S3FileTransfer) listObjectsWithPrefix(
+	bucketName string,
+	prefix string,
+) ([]*s3.GetObjectInput, error) {
+	var objects []*s3.GetObjectInput
+	params := &s3.ListObjectsV2Input{
+		Bucket: aws.String(bucketName),
+		Prefix: aws.String(prefix),
+	}
+	isTruncated := true
+	for isTruncated {
+		output, err := ft.client.ListObjectsV2(ft.ctx, params)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, object := range output.Contents {
+			objects = append(objects, &s3.GetObjectInput{
+				Bucket: aws.String(bucketName),
+				Key:    object.Key,
+			})
+		}
+
+		isTruncated = *output.IsTruncated
+		if isTruncated {
+			params.ContinuationToken = output.NextContinuationToken
+		}
```

Comment:
question/suggestion: might be off, but just in case -- does this need to be checked for `nil`?
```suggestion
		if output.IsTruncated != nil {
		    isTruncated = *output.IsTruncated
		} else {
		    isTruncated = false
		}
		if isTruncated {
			params.ContinuationToken = output.NextContinuationToken
		}
```

Reply:
good point! I want to say that we should be able to rely on this not being `nil` based on the docs, but I think its better to be safe since s3 doesn't explicitly guarantee that.